_posts/2019-09-29-Large-NLP-in-prod.md

```diff
@@ -0,0 +1,112 @@
+
+Part of the latter is something along the lines of:
+
+```python
+from pytorch_pretrained_bert import BertForSequenceClassification
+
+def load_model(model_dir: str, num_labels: int):
+    model = BertForSequenceClassification.from_pretrained(
+        model_dir,
+        num_labels=num_labels
+    )
+    model.eval()
+    return model
+``` 
+
+Once these steps have finished, the service will start serving requests. Each of these services will have an endpoint (or [Sanic route](https://sanic.readthedocs.io/en/latest/sanic/routing.html)) to get the model's predictions for a given input - it broadly looks like this:
+
+```python
+@app.route("/predict/", methods=['POST'])
+async def predict(request):
+    text = request.json["text"]
+    # Run some validations or do other checks
+    label, probability = await model_predict(text)
+    return response.json(
+        {
+            "label": label,
+            "probability": probability,
+            # And some other fields
+        },
+        status=200
+    )
+```
+
+The `async` and `await` syntax is the key here: handlers are an [async co-routine](https://docs.python.org/3/library/asyncio-task.html), and all model predictions are run in the `asyncio` event loop, rather than as blocking functions. There are many blog posts that describe the details of the Python event loop and the `async / await` syntax much better than I ever could; if you're interested, I'd recommend searching for this topic ([here's one](https://stackabuse.com/python-async-await-tutorial/)).
+
+The `model_predict()` runs model predictions with `torch.no_grad()`: this ensures that the autograd engine is not used. According to [this thread](https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615), this reduces memory usage and speeds up computation. Naturally, you can't then backpropate (needing to do so while serving is an entirely different type of problem).
+
+There was one tiny trick that two folks on the team discovered which helped us to make these types of services _even faster_. They discovered the issue that it seems [others have also found](https://twitter.com/MarkNeumannnn/status/1067926695338926080) regarding threading performance and the `OMP_NUM_THREADS` and `MKL_NUM_THREADS` environment variables; the one difference was that they also had to factor in how all of this plays with Sanic [worker threads](https://sanic.readthedocs.io/en/latest/sanic/deploying.html#workers).
+
+## 🔍 Closing thoughts
+
+The main disclaimer that I'll add to the above is that we are currently in the stage of development where we're iterating on and validating new product features, and not squeezing performance gains out of existing ones. Perhaps, once we reach that stage (where every microsecond matters), I'll change my mind about BERT being suitable for production 😊.
```

Comment:
I guess for us, it's less about the microseconds. It's more that each instance needs to take up, say, 500mb of memory to serve a single model. If we need to scale these services quite often due their predictions taking a few hundred milliseconds, and thus, only being able to run a few predictions per second through each instance, we then end up needing to run many, many instances: which can end up in 5-10GB of memory used across the cluster. A normal Go binary can run in probably 100x less memory.

It's not a real issue - servers are pretty cheap, all things given - but just something worth bearing in mind at a larger scale. I think you've covered it pretty well.